crypto/threads_pthread.c

```diff
@@ -42,12 +44,481 @@
+        qp_idx = (uint64_t)__ATOMIC_LOAD_N(&ilock->reader_idx, __ATOMIC_SEQ_CST);
+        __ATOMIC_ADD_FETCH(&ilock->qp_group[qp_idx].users, VAL_READER,
+                           __ATOMIC_SEQ_CST);
+        /* if the idx hasn't changed, we're good, else try again */
+        if (qp_idx == (uint64_t)__ATOMIC_LOAD_N(&ilock->reader_idx, __ATOMIC_SEQ_CST))
+            break;
+
+        __ATOMIC_SUB_FETCH(&ilock->qp_group[qp_idx].users, VAL_READER,
+                           __ATOMIC_SEQ_CST);
+    }
+
+    return (struct rcu_qp *)&ilock->qp_group[qp_idx];
+}
+
+/* Public READ lock api */
+void ossl_rcu_read_lock(CRYPTO_RCU_LOCK *lock)
+{
+    struct rcu_thr_data *data;
+
+    /*
+     * we're going to access current_qp here so ask the
+     * processor to fetch it
+     */
+    data = CRYPTO_THREAD_get_local(&rcu_thr_key);
+
+    if (data == NULL) {
+        for (;;) {
+            data = OPENSSL_zalloc(sizeof(struct rcu_thr_data));
+            if (data != NULL)
+                break;
+        }
+        CRYPTO_THREAD_set_local(&rcu_thr_key, data);
+        ossl_init_thread_start(NULL, NULL, free_rcu_thr_data);
+    }
+
+    if (data->qp == NULL)
+        data->qp = get_hold_current_qp(lock);
+
+    /* inc our local count */
+    data->count++;
+
+    /* check for underflow condition */
+    assert(data->count >= 0);
+}
+
+/* Public READ unlock api */
+void ossl_rcu_read_unlock(CRYPTO_RCU_LOCK *lock)
+{
+    struct rcu_thr_data *data = CRYPTO_THREAD_get_local(&rcu_thr_key);
+
+    assert(data != NULL);
+
+    data->count--;
+
+    assert(data->count >= 0);
+
+    if (data->count == 0) {
+        __ATOMIC_SUB_FETCH(&data->qp->users, VAL_READER,
+                           __ATOMIC_SEQ_CST);
+        data->qp = NULL;
+    }
+}
+
+/*
+ * Write side allocation routine to get the current qp
+ * and replace it with a new one
+ */
+static struct rcu_qp *update_qp(struct rcu_lock_internal *ilock)
+{
+    uint64_t new_id;
+    uint64_t current_idx;
+
+    pthread_mutex_lock(&ilock->alloc_lock);
+
+    /*
+     * we need at least one qp to be available with one
+     * left over, so that readers can start working on
+     * one that isn't yet being waited on
+     */
+    while (ilock->group_count - ilock->writers_alloced < 2) {
+        /* we have to wait for one to be free */
+        pthread_cond_wait(&ilock->alloc_signal, &ilock->alloc_lock);
+    }
+
+    current_idx = ilock->current_alloc_idx;
+
+    /* Allocate the qp */
+    ilock->writers_alloced++;
+
+    /* increment the allocation index */
+    ilock->current_alloc_idx =
+        (ilock->current_alloc_idx + 1) % ilock->group_count;
+
+    /* get and insert a new id */
+    new_id = __ATOMIC_FETCH_ADD(&ilock->id_ctr, 1, __ATOMIC_SEQ_CST);
+
+    new_id = VAL_ID(new_id);
+    __ATOMIC_AND_FETCH(&ilock->qp_group[current_idx].users, ID_MASK,
+                       __ATOMIC_SEQ_CST);
+    __ATOMIC_OR_FETCH(&ilock->qp_group[current_idx].users, new_id,
+                      __ATOMIC_SEQ_CST);
```

Comment:
its still strictly speaking atomic, since this is protected by the write alloc lock on the write side.  I used the atomics here to be consistent with the atomic usage of the variables on the read side.  It could be replaced with a straight bitwise or/and operation, but I think that would look confusing to readers.

Reply:
I'd use normal operations and add a comment. No need to use atomics where they aren't necessary.